### PR TITLE
Move setjmp/longjmp helper functions and globals to compiler_rt_wasm library

### PR DESCRIPTION
**Pull request by sbc100**
Previously these have been generated by LLVM during codegen
but this doesn't work once we have separate compilation.

See LLVM change: https://reviews.llvm.org/D49208

### CODE REVIEW

**Comment by sbc100** [comment]
Currently these two functions for force-included in the link.  Not sure if that can be avoided somehow?

File: emcc.py [hunk 1]
```diff
@@ -1120,6 +1120,10 @@ def check(input_file):
         shared.Settings.EXPORTED_FUNCTIONS += ['___cxa_demangle']
         forced_stdlibs += ['libcxxabi']
 
+      # These is needed by setjmp as well as C++ esceptions.
```
Comment (aheejin) [inline emcc.py]: exceptions

File: emscripten.py [hunk 2]
```diff
@@ -117,7 +117,7 @@ def compile_js(infile, temp_files, DEBUG):
   return backend_output
 
 
-def parse_backend_output(backend_output, DEBUG):
+def parse_fastcomp_output(backend_output, DEBUG):
```
Comment (aheejin) [inline emscripten.py]: This is used by the wasm backend as well. why rename to fastcomp?

File: system/lib/compiler-rt/extras.c [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    __threwValue = value;
+  }
+}
+
+void setTempRet0(int value) {
+  __tempRet0 = value;
+}
+
+void getTempRet0(int value) {
+  __tempRet0 = value;
```
Comment (aheejin) [inline system/lib/compiler-rt/extras.c]: This is the same as setTempRet0. Should be something like
```
int getTempRet0() {
  return __tempRet0;
}
```

By the way, do we need this? I don't think we generated `getTempRet0` in the LLVM pass or needed it.

File: tools/shared.py [hunk 4]
```diff
@@ -1924,6 +1924,10 @@ def wrapped():
     else:
       for export in expand_response(Settings.EXPORTED_FUNCTIONS):
         cmd += ['--export', export[1:]] # Strip the leading underscore
+        # For certainly symbols we also need to use --undefined so that the
```
Comment (aheejin) [inline tools/shared.py]: What are 'certainly symbols'?

**Comment by aheejin** [comment]
And for the title.. these are not only used for sjlj but also for EH in general.

File: emcc.py (hunk 1 above)
Comment (kripken) [inline emcc.py]: `is` => `are`

File: emcc.py [hunk 5]
```diff
@@ -1120,6 +1120,10 @@ def check(input_file):
...
+      # TODO(sbc): Do this conditionally
```
Comment (kripken) [inline emcc.py]: for now, how about doing this only for the wasm backend?

File: system/lib/compiler-rt/extras.c [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// Supprot functions for emscripten setjmp/longjmp and excecption handling
```
Comment (kripken) [inline system/lib/compiler-rt/extras.c]: `supprot` => `support`

File: system/lib/compiler-rt/extras.c (hunk 3 above)
Comment (jgravelle-google) [reply system/lib/compiler-rt/extras.c]: We've generated or needed it before for runtime compatibility reasons... I thought we did this in binaryen in wasm-emscripten.cpp, but I don't see it anywhere. Not even in s2wasm from back when that still existed.
So maybe we don't need this? Somehow?

File: tools/shared.py [hunk 7]
```diff
@@ -1924,6 +1924,10 @@ def wrapped():
...
+        # symbol is required to exist and will be pulled from .a file if needed.
```
Comment (jgravelle-google) [inline tools/shared.py]: nit: "from a .a file" maybe?

**Comment by sbc100** [comment]
Looks like the weak linking solution will work for now so abandoning this change.

Its not ideal since every object will end up with the copy of the same 5 symbols, which is kind of ugly, but its the smallest incremental change right now.

**Comment by dschuff** [comment]
Oh, sorry I commented on https://reviews.llvm.org/D49263 because I saw it first, but I'll add it here since we have all the stakeholders on this PR already. 
I can see how this approach requires less code, but I think I actually prefer the approach of linking this code in from a library rather than generating N copies of it. It's more consistent with how system code is usually included into programs (i.e. linking it in rather than generating it magically from the compiler). It's also easier to read and understand for non-compiler developers (and for compiler developers who can't remember exactly where in the compiler to look :) ).

**Pull request closed without merging**